File: Chapter13/VB/Client Side/Order Entry/FormMain.vb
Public Class FormMain
  Inherits System.Windows.Forms.Form

  Private _workerOrderEntry As WorkerOrderEntry
  Private _coordinatorOrderEntry As CoordinatorOrderEntry
  Private _router As Router

  Sub Build()
    _workerOrderEntry = New WorkerOrderEntry
    _coordinatorOrderEntry = New CoordinatorOrderEntry
    _router = New Router
  End Sub

  Sub Bind()
    _workerOrderEntry.OnGetColors = AddressOf _coordinatorOrderEntry.GetColors 'delegate
    _workerOrderEntry.OnGetOptions = AddressOf _coordinatorOrderEntry.GetOptions 'delegate
    _workerOrderEntry.OnGetStyles = AddressOf _coordinatorOrderEntry.GetStyles 'delegate
    AddHandler _workerOrderEntry.OnSubmitOrder, AddressOf _coordinatorOrderEntry.SubmitOrder 'event
    _workerOrderEntry.OnComputeCostOfOptions = AddressOf _coordinatorOrderEntry.ComputeCostOfOptions 'delegate

    _coordinatorOrderEntry.OnGetColors = AddressOf _router.GetColors 'delegate
    _coordinatorOrderEntry.OnGetOptions = AddressOf _router.GetOptions 'delegate
    _coordinatorOrderEntry.OnGetStyles = AddressOf _router.GetStyles 'delegate
    AddHandler _coordinatorOrderEntry.OnSubmit, AddressOf _router.SubmitOrder 'event
  End Sub

  Sub SetupUi()
    Controls.Add(_workerOrderEntry.PanelMain)
    PopulateModels()
  End Sub

  Sub PopulateModels()
    Try
      Dim models As String() = _router.GetModels()
      _workerOrderEntry.PopulateModels(models)
    Catch ex As Exception
      Dim msg As String = String.Format("The server must be running before starting the Order Entry program. " + _
                                  "To run the server, start OrderProcessingHostProgram.exe.\nException Details: [{0}]", ex.Message)
      MessageBox.Show(msg, "An exception occurred while connecting to server")
      Throw ex
    End Try
  End Sub

#Region " Windows Form Designer generated code "

  Public Sub New()
    MyBase.New()

    'This call is required by the Windows Form Designer.
    InitializeComponent()

    Build()
    Bind()

    SetupUi()
  End Sub

  'Form overrides dispose to clean up the component list.
  Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
    If disposing Then
      If Not (components Is Nothing) Then
        components.Dispose()
      End If
    End If
    MyBase.Dispose(disposing)
  End Sub

  'Required by the Windows Form Designer
  Private components As System.ComponentModel.IContainer

  'NOTE: The following procedure is required by the Windows Form Designer
  'It can be modified using the Windows Form Designer.  
  'Do not modify it using the code editor.
  <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()
    '
    'Form1
    '
    Me.AutoScaleBaseSize = New System.Drawing.Size(6, 15)
    Me.ClientSize = New System.Drawing.Size(468, 387)
    Me.Name = "Form1"
    Me.StartPosition = System.Windows.Forms.FormStartPosition.CenterScreen
    Me.Text = "ASAP Cars - Custom Tailor Your Car"

  End Sub

#End Region

  Shared Sub Main()
    Dim _formMain As FormMain = Nothing
    Try
      _formMain = New FormMain
    Catch ex As Exception
      ' exceptions typically occur if the client can't connect to the server
      MessageBox.Show(ex.Message, "An exception occurred")
      Return
    End Try

    Application.Run(_formMain)
  End Sub

End Class

File: Chapter13/C#/Client Side/Order Entry/FormMain.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;

namespace Asap.Cars.OrderEntry
{
	/// <summary>
	/// Summary description for FormMain.
	/// </summary>
  public class FormMain : System.Windows.Forms.Form
  {
	  /// <summary>
	  /// Required designer variable.
	  /// </summary>
	  private System.ComponentModel.Container components = null;

    WorkerOrderEntry workerOrderEntry;
    CoordinatorOrderEntry coordinatorOrderEntry;
    Router router;

	  public FormMain()
	  {
		  //
		  // Required for Windows Form Designer support
		  //
		  InitializeComponent();

      Build();
      Bind();

      SetupUi();
	  }

    void Build()
    {
      workerOrderEntry = new WorkerOrderEntry();
      coordinatorOrderEntry = new CoordinatorOrderEntry();
      router = new Router();
    }

    void Bind()
    {
      workerOrderEntry.OnGetColors += new Asap.Cars.OrderEntry.WorkerOrderEntry.GetColorsHandler(coordinatorOrderEntry.GetColors);
      workerOrderEntry.OnGetOptions += new Asap.Cars.OrderEntry.WorkerOrderEntry.GetOptionsHandler(coordinatorOrderEntry.GetOptions);
      workerOrderEntry.OnGetStyles += new Asap.Cars.OrderEntry.WorkerOrderEntry.GetStylesHandler(coordinatorOrderEntry.GetStyles);
      workerOrderEntry.OnSubmitOrder += new Asap.Cars.OrderEntry.WorkerOrderEntry.SubmitOrderHandler(coordinatorOrderEntry.SubmitOrder);
      workerOrderEntry.OnComputeCostOfOptions += new Asap.Cars.OrderEntry.WorkerOrderEntry.ComputeCostOfOptionsHandler(coordinatorOrderEntry.ComputeCostOfOptions);

      coordinatorOrderEntry.OnGetColors += new Asap.Cars.OrderEntry.CoordinatorOrderEntry.GetColorsHandler(router.GetColors);
      coordinatorOrderEntry.OnGetOptions += new Asap.Cars.OrderEntry.CoordinatorOrderEntry.GetOptionsHandler(router.GetOptions);
      coordinatorOrderEntry.OnGetStyles += new Asap.Cars.OrderEntry.CoordinatorOrderEntry.GetStylesHandler(router.GetStyles);
      coordinatorOrderEntry.OnSubmit += new Asap.Cars.OrderEntry.CoordinatorOrderEntry.SubmitHandler(router.SubmitOrder);
    }

    void SetupUi()
    {
      Controls.Add(workerOrderEntry.panelMain);
      PopulateModels();
    }

    void PopulateModels()
    {
      try
      {
        string[] models = router.GetModels();
        workerOrderEntry.PopulateModels(models);
      }
      catch (Exception ex)
      {
        string msg = string.Format("The server must be running before starting the Order Entry program. " +
                                    "To run the server, start OrderProcessingHostProgram.exe.\nException Details: [{0}]", ex.Message);
        MessageBox.Show(msg, "An exception occurred while connecting to server");
        throw ex;
      }
    }


	  /// <summary>
	  /// Clean up any resources being used.
	  /// </summary>
	  protected override void Dispose( bool disposing )
	  {
		  if( disposing )
		  {
			  if (components != null) 
			  {
				  components.Dispose();
			  }
		  }
		  base.Dispose( disposing );
	  }

	  #region Windows Form Designer generated code
	  /// <summary>
	  /// Required method for Designer support - do not modify
	  /// the contents of this method with the code editor.
	  /// </summary>
	  private void InitializeComponent()
	  {
      // 
      // Form1
      // 
      this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
      this.ClientSize = new System.Drawing.Size(390, 336);
      this.Name = "Form1";
      this.StartPosition = System.Windows.Forms.FormStartPosition.CenterScreen;
      this.Text = "ASAP Cars - Custom Tailor Your Car";

    }
	  #endregion

	  /// <summary>
	  /// The main entry point for the application.
	  /// </summary>
	  [STAThread]
	  static void Main() 
	  {
      FormMain formMain = null;
      try
      {
        formMain = new FormMain();
      }
      catch (Exception ex)
      {
        // exceptions typically occur if the client can't connect to the server
        MessageBox.Show(ex.Message, "An exception occurred");
        return;
      }

		  Application.Run(formMain);
	  }

  }
}
